text/0000-postfix-match.md

```diff
@@ -0,0 +1,321 @@
+```
+
+## async support
+
+One thing of note is that in option chains you cannot use futures unless you use adapters
+like [`OptionFuture`](https://docs.rs/futures/0.3.21/futures/future/struct.OptionFuture.html).
+
+Using match, you can avoid that by supporting `.await` directly.
+
+```rust
+context.client
+    .post("https://example.com/crabs")
+    .body("favourite crab?")
+    .send()
+    .await
+    .match {
+        Err(_) => Ok("Ferris"),
+        Ok(resp) => resp.json::<String>().await,
+        //         this works in a postfix-match ^^^^^^
+    }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`X.Y.match {}.Z`
+
+Will be interpreted as
+
+```rust
+match X.Y {}.Z`
+```
+
+I believe this would be the same precedence as `await`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Having multiple forms of match could be confusing. However, I believe
+that most issues could be resolved if clippy would detect uses of
+
+```rust
+x.match {}
+```
+
+when the desugared
+
+```rust
+match x {}
+```
+
+would fit on a single line after being rustfmt. The opposite could
+also be true - if the scrutinee spans multiple lines, it should be made into
+a postfix form.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Rationale
+
+The core rationale is that a lot of these method chain functions are designed to
+avoid using bulky match statements that ruin the flow.
+
+Rather than keep adding more of these methods to suit the needs, we should
+make the language more flexible such that match statements aren't a hindrance.
+
+## Alternatives
+
+### Postfix Macros
+
+[postfix macros](https://github.com/rust-lang/rfcs/pull/2442) have been an idea for many years now.
+If they were to land, this feature could easily be implemented as a macro:
+
+```rust
+macro match_! (
+    { $self:expr; $(
+        $arm:pat => $body:expr,
+    )+ } => {
+        match $self { $(
+            $arm => $body,
+        )+ }
+    }
+)
+```
+
+However, after years of discussion and hundreds of thumbs-ups, it feels like we're
+still not close to agreeing on syntax or behaviour.
+
+### Pipes
+
+I've already mentioned [tap](#tap-pipelines) for how we can do prefix-in-postfix,
+so we could promote the use of `.pipe()` instead. However, using the pipe method makes
+async awkward and control flow impossible.
+
+```rust
+x.pipe(|x| async {
+    match x {
+        Err(_) => Ok("Ferris"),
+        Ok(resp) => resp.json::<String>().await,
+    }
+}).await
+```
+
+We could add a new builtin pipeline operator
+(e.g. `|>`<sup>[0]</sup> or `.let x {}`><sup>[1]</sup>)
+but this is brand new syntax and requires a whole new set of discussions.
+
+[0]: https://internals.rust-lang.org/t/idea-operator-for-postfix-calls-to-macros-and-free-functions/7634
+[1]: https://internals.rust-lang.org/t/idea-postfix-let-as-basis-for-postfix-macros/12438
+
+# Prior art
+[prior-art]: #prior-art
```

Comment:
Something else you could add here:

C# has had C- and Java-style `switch (foo) { … }` since the very beginning.  But [C# 8.0 added postfix `switch` expressions](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#switch-expressions) that use `foo switch { … }` instead.  That's very much like how Rust has had `match foo { … }` since the beginning while this RFC proposes adding `foo.match { … }`.

It's fun to see just how rusty that feature is.  To use the example from the [`csharp_style_prefer_switch_expression` style rule](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0066) (roughly the equivalent of a warn-by-default clippy lint), instead of 
```cs
switch (x)
{
    case 1:
        return 1 * 1;
    case 2:
        return 2 * 2;
    default:
        return 0;
}
```

It allows doing
```cs
return x switch
{
    1 => 1 * 1,
    2 => 2 * 2,
    _ => 0,
};
```
Using `=>` and `_` *exactly* the same way Rust does here.  (Not that Rust invented it; it's also the same as many other languages.)

This isn't a perfect analog, since C#'s change does include a bunch of pattern syntax cleanup.  But it's still another language in which it was decided to add a postfix matching operator even though everything it does could still be done (sometimes less nicely) with the existing prefix matching operator.

Reply:
Ah yes of course! I even use them all the time at work 😅

Reply:
In C# the non-postfix `switch` is not an expression, in Rust it is.